internal/utils/utils_test.go

```diff
@@ -172,3 +178,81 @@ var _ = Describe("InheritNotReadyConditionFromObjects", func() {
 		Expect(readyCondition.Message).To(Equal(`CommitStatus "child1" is not Ready because "NotReady1": Child1 is not ready`))
 	})
 })
+
+var _ = Describe("HandleReconciliationResult panic recovery", func() {
+	var (
+		ctx      context.Context
+		obj      *promoterv1alpha1.PromotionStrategy
+		recorder record.EventRecorder
+		scheme   *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+		obj = &promoterv1alpha1.PromotionStrategy{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "PromotionStrategy",
+				APIVersion: "promoter.argoproj.io/v1alpha1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       "test-strategy",
+				Namespace:  "default",
+				Generation: 1,
+			},
+		}
+		scheme = runtime.NewScheme()
+		_ = promoterv1alpha1.AddToScheme(scheme)
+		recorder = record.NewFakeRecorder(10)
+	})
+
+	It("should recover from panic and convert it to an error", func() {
+		var err error
+		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithStatusSubresource(obj).Build()
+
+		// This function will panic, and HandleReconciliationResult should recover from it
+		func() {
+			defer utils.HandleReconciliationResult(ctx, metav1.Now().Time, obj, fakeClient, recorder, &err)
+			panic("test panic message")
+		}()
+
+		// The panic should have been caught and converted to an error
+		Expect(err).To(HaveOccurred())
+		Expect(err.Error()).To(ContainSubstring("panic in reconciliation"))
+		Expect(err.Error()).To(ContainSubstring("test panic message"))
+	})
+
+	It("should handle normal errors without panicking", func() {
+		var err error
+		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithStatusSubresource(obj).Build()
+
+		// Create the object in the fake client so HandleReconciliationResult can update it
+		Expect(fakeClient.Create(ctx, obj)).To(Succeed())
+
+		// This function will return an error normally
+		func() {
+			defer utils.HandleReconciliationResult(ctx, metav1.Now().Time, obj, fakeClient, recorder, &err)
+			err = errors.New("test error message")
+		}()
+
+		// The error should be preserved
+		Expect(err).To(HaveOccurred())
+		Expect(err.Error()).To(ContainSubstring("test error message"))
+	})
+
+	It("should handle successful reconciliation without error", func() {
+		var err error
+		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithStatusSubresource(obj).Build()
```

Comment:
```suggestion
		// We use fakeclient here since it's virtually impossible to trigger a panic otherwise. Don't spread
		// this use to other tests if at all avoidable.
		fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithStatusSubresource(obj).Build()
```